keps/sig-apps/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1250 @@
+They can only expand the volumes, or modify them with VolumeAttributesClass
+by updating individual PersistentVolumeClaim objects as an ad-hoc operation.
+When the StatefulSet scales up, the new PVC(s) will be created with the old
+config and this again needs manual intervention.
+This brings many headaches in a continuously evolving environment.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+* Allow users to update some fields of `volumeClaimTemplates` of a `StatefulSet`, specifically:
+  * increasing the requested storage size (`spec.volumeClaimTemplates.spec.resources.requests.storage`)
+  * modifying Volume AttributesClass used by the claim( `spec.volumeClaimTemplates.spec.volumeAttributesClassName`)
+  * modifying VolumeClaim template's labels (`spec.volumeClaimTemplates.metadata.labels`)
+  * modifying VolumeClaim template's annotations (`spec.volumeClaimTemplates.metadata.annotations`)
+* Add `.spec.volumeClaimUpdatePolicy` allowing users to decide how the volume claim will be updated: in-place or on PVC deletion.
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+* Support automatic re-creating of PersistentVolumeClaim. We will never delete a PVC automatically.
+* Validate the updated `volumeClaimTemplates` as how PVC patch does.
+* Update ephemeral volumes.
+* Patch PVCs that are different from the template, e.g. StatefulSet adopts the pre-existing PVCs.
+* Support for volumes that only support offline expansion.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Kubernetes API Changes  
+
+Change API server to allow specific updates to `volumeClaimTemplates` of a StatefulSet:
+   * `spec.volumeClaimTemplates.spec.resources.requests.storage` (increase only)
+   * `spec.volumeClaimTemplates.spec.volumeAttributesClassName`
+   * `spec.volumeClaimTemplates.metadata.labels`
+   * `spec.volumeClaimTemplates.metadata.annotations`
+
+Introduce a new field in StatefulSet `spec`: `volumeClaimUpdatePolicy` to
+specify how to coordinate the update of PVCs and Pods. Possible values are:
+- `OnClaimDelete`: the default value, only update the PVC when the the old PVC is deleted.
+- `InPlace`: patch the PVC in-place if possible. Also includes the `OnClaimDelete` behavior.
+
+
+Additionally collect the status of managed PVCs, and show them in the StatefulSet status.
+Some fields in the `status` are updated to reflect the status of the PVCs:
+- claimsReadyReplicas: the number of replicas with all PersistentVolumeClaims ready to use.
```

Comment:
This is a new field, right?

Reply:
As mentioned in other comments, I'm against adding one more readyreplicas field which will be very confusing to users.

Reply:
I think we will need a field to expose the PVC readiness, for `kubectl rollout status` to wait for the last replica to be ready before returning. Any idea better than `claimsReadyReplicas`?